File: projects/barf/barf/util/render.py
"""Rendering network.yml hosts to device configs.

Shared by ``barf config generate|diff|deploy``.
"""

import os
from typing import List

import yaml

from barf.common import render_template
from barf.configs import BLOCK_VENDORS, build_context, render_blocks
from barf.model.wireguard import WGNetworkLink, prefetch_keypairs
from barf.util.sites import SITE_ORIGIN_FUNC
from barf.vendors import BaseHost


def prefetch_link_keys(targets: List[BaseHost], links: List[WGNetworkLink]) -> None:
    """Warm the WireGuard key cache for every link touching ``targets``.

    Rendering a host needs the keypairs of BOTH sides of each of its
    links; fetching them serially from Vault dominates render time, so
    pull them concurrently up front.
    """
    target_names = {host.hostname for host in targets}
    paths = set()
    for link in links:
        # IPsec links have no WireGuard keypairs in Vault.
        if link.ipsec:
            continue
        if link.side_a.hostname in target_names or link.side_b.hostname in target_names:
            paths.add(link._key_path(link.side_a.hostname))
            paths.add(link._key_path(link.side_b.hostname))

    prefetch_keypairs(sorted(paths))


def render_host_config(
    host: BaseHost,
    links: List[WGNetworkLink],
    global_meta: dict,
    secrets,
) -> str:
    """Render the config for one network.yml host.

    Args:
        host: The host to render.
        links: All network links; the host's VPN links are selected here.
        global_meta: The network.yml ``global_meta`` block.
        secrets: The Vault secret fetcher passed through to templates.
    """
    ctx = build_context(host, links, global_meta, secrets)

    if host.devicetype in BLOCK_VENDORS:
        return render_blocks(ctx)

    role_meta = {}
    if host.role == "vpn":
        role_meta = {
            "vpn_links": ctx.vpn_links,
            "site_import_rules": ctx.site_import_rules,
            "origin_site": ctx.origin_site,
            "community_asn": ctx.community_asn,
            "SITE_ORIGIN_FUNC": SITE_ORIGIN_FUNC,
        }

    return render_template(
        f"{host.role}/{host.devicetype}.j2",
        device=host,
        secrets=secrets,
        global_meta=global_meta,
        **role_meta,
    )


def write_rendered_config(
    host: BaseHost, rendered: str, output_dir: str = "output"
) -> str:
    """Write a rendered config (and its cloud-init twin) under ``output_dir``.

    Returns:
        The path of the main config file.
    """
    role_dir = os.path.join(output_dir, host.role)
    os.makedirs(os.path.join(role_dir, "cloud_init"), exist_ok=True)

    config_path = os.path.join(role_dir, host.hostname)
    with open(config_path, "w") as f:
        f.write(rendered)

    with open(os.path.join(role_dir, "cloud_init", host.hostname), "w") as f:
        f.write("#cloud-config\n")
        yaml.dump(
            {"vyos_config_commands": [x for x in rendered.split("\n") if x]},
            f,
        )

    return config_path

File: go/pkg/barf/render/render.go
// Package render turns a parsed network.yml host into device config text.
//
// It is a port of the Python barf render path (barf/util/render.py plus
// barf/configs/): one block per feature, one method per vendor, emitting
// the exact strings the Python implementation emits. The goldens in
// projects/barf/tests/golden/ are the byte-for-byte parity contract.
//
// Nothing here talks to a device or to Vault: secrets arrive through the
// SecretSource interfaces below.
package render

import (
	"fmt"

	"github.com/general-programming/megarepo/go/pkg/barf/model"
)

// Renderer produces the device config text for one host.
type Renderer interface {
	Render(h *model.Host, n *model.Network, s SecretSource) (string, error)
}

// SecretSource resolves per-host secrets (Vault in production, a
// deterministic fake in tests). Mirrors Python BaseHost.secret().
type SecretSource interface {
	HostSecret(hostname, key string) (string, error)
}

// VaultSource resolves the fabric-wide Vault attribute lookups the
// Python templates reach through util.secrets.VaultSecrets (the VyOS API
// key, IPsec pre-shared secrets). Optional: a SecretSource that does not
// implement it makes those renders fail loudly rather than silently.
type VaultSource interface {
	VaultSecret(key string) (string, error)
}

// TacacsSource resolves a host's TACACS+ shared key. Python reads it
// from a fleet-wide secret (cluster-secrets/tacacs-keys, field
// <hostname>, subkey "key") rather than the per-host path, so it does
// not fit HostSecret's (hostname, key) shape. Optional, like
// VaultSource: only the IOS-family vendors need it.
type TacacsSource interface {
	TacacsKey(hostname string) (string, error)
}

// Keypair is one WireGuard keypair, base64 as stored in Vault.
type Keypair struct {
	Public  string
	Private string
}

// WireguardSource resolves the fabric mesh keypairs by Vault path
// (model.Link.KeyPath). Optional, like VaultSource.
type WireguardSource interface {
	WireguardKeypair(path string) (Keypair, error)
}

// renderers is the devicetype -> Renderer registry, mirroring the
// Python VENDOR_MAP. Every vendor Python renders config for is here;
// `external` deliberately is not (see Templatable).
//
// Which ROLES each vendor supports is the renderer's own business, and
// the answer is narrow, because Python's role dispatch is just
// `templates/<role>/<devicetype>.j2` existing on disk:
//
//	vyos, linux, mikrotik   vpn only (barf.configs.BLOCK_REGISTRY)
//	edgeos                  vpn only (templates/vpn/edgeos.j2)
//	cisco, dnos6, dnos9     network_devices only
//	eos                     any role -- arista.py renders a scoped
//	                        managed slice, which render_host_config
//	                        dispatches to before role is consulted
//
// Those are all the (role, devicetype) pairs that exist: there is no
// templates/vpn/vyos.j2, no templates/core/anything. A pair outside the
// table fails in Python with a Jinja TemplateNotFound and fails here
// with noTemplateError, which is the same answer said politely.
var renderers = map[string]Renderer{
	"eos":      EOS{},
	"vyos":     VyOS{},
	"linux":    Linux{},
	"mikrotik": MikroTik{},
	"edgeos":   EdgeOS{},
	"cisco":    Cisco{},
	"dnos6":    DNOS6{},
	"dnos9":    DNOS9{},
}

// aliases are the extra VENDOR_MAP spellings, which come from NetBox
// platform slugs rather than from network.yml.
var aliases = map[string]string{
	"cisco-ios": "cisco",
	"dnos-6":    "dnos6",
	"dnos-9":    "dnos9",
}

// Templatable reports whether barf generates config for a device type at
// all (Python BaseHost.TEMPLATABLE).
//
// Only `external` is not: it marks the far side of a fabric link that
// barf does not manage (a cloud VPN endpoint, a peer's router). Such
// hosts exist in network.yml so links can name them and so their ASN
// and endpoint address are known, but nothing is ever rendered FOR
// them. Callers walking the fleet should skip them by asking this,
// rather than treating them as a vendor whose port has not landed.
func Templatable(deviceType string) bool {
	return deviceType != "external"
}

// For returns the renderer registered for a device type.
func For(deviceType string) (Renderer, bool) {
	if alias, ok := aliases[deviceType]; ok {
		deviceType = alias
	}
	renderer, ok := renderers[deviceType]
	return renderer, ok
}

// Host renders one host with the renderer registered for its device type.
func Host(h *model.Host, n *model.Network, s SecretSource) (string, error) {
	if !Templatable(h.DeviceType) {
		return "", fmt.Errorf("%s: device type %q is unmanaged; barf renders no config for it",
			h.Hostname, h.DeviceType)
	}
	renderer, ok := For(h.DeviceType)
	if !ok {
		return "", fmt.Errorf("%s: no renderer for device type %q", h.Hostname, h.DeviceType)
	}
	return renderer.Render(h, n, s)
}

// noTemplateError is the Go spelling of Jinja's TemplateNotFound for a
// (role, devicetype) pair barf has no config for.
func noTemplateError(h *model.Host) error {
	return fmt.Errorf("%s: no config for role %q on device type %q",
		h.Hostname, h.Role, h.DeviceType)
}

func vaultSecret(s SecretSource, key string) (string, error) {
	source, ok := s.(VaultSource)
	if !ok {
		return "", fmt.Errorf("secret source cannot resolve vault key %q", key)
	}
	return source.VaultSecret(key)
}

func tacacsKey(s SecretSource, hostname string) (string, error) {
	source, ok := s.(TacacsSource)
	if !ok {
		return "", fmt.Errorf("secret source cannot resolve the tacacs key for %q", hostname)
	}
	return source.TacacsKey(hostname)
}

func wireguardKeypair(s SecretSource, path string) (Keypair, error) {
	source, ok := s.(WireguardSource)
	if !ok {
		return Keypair{}, fmt.Errorf("secret source cannot resolve wireguard keypair %q", path)
	}
	return source.WireguardKeypair(path)
}
